### DISCUSSION TITLE
File default constructor

### DISCUSSION THREAD

**Discussion by gouarin** [BlueBrain/HighFive#756] [Questions]
Hi,
I would like to create a c++ class with a `HighFive::File` attribute. This attribute is created in a different way depending on the options given by the user: it can be an hdf5 file with MPIOFileAccess property or each process opens its own hdf5 file.
Unfortunately, it's not possible to do that because the default `File` constructor is protected. Could you explain why you made this choice?
Thanks

**Comment by 1uc** [comment]
A typical way to get around this type of limitation is:
```
class ObjWithFile {
public:
  ObjWithFile(...) : _file(open_file(...)) {}

protected:
  static HighFive::File open_file(...) {
    if(...) {
      return HighFive::File(...);
    } else {
      return HighFive::File(...);
    }
private:
  HighFive::File _file;
};
```
The choice is between preventing incorrectly initialized objects; and programming convenience in cases such as the one you describe.

**Reply by gouarin** [reply]
Thanks @1uc for your quick reply and the tip!
